fuel-merkle/src/binary/in_memory.rs

```diff
@@ -62,6 +67,54 @@ impl Default for MerkleTree {
     }
 }
 
+#[derive(Default, Debug, Clone)]
+pub struct MerkleRootCalculator {
+    stack: Vec<Node>,
+}
+
+impl MerkleRootCalculator {
+    pub fn new() -> Self {
+        Self { stack: Vec::new() }
+    }
+
+    pub fn push(&mut self, data: &[u8]) {
+        let node = Node::create_leaf(0, data);
+        self.stack.push(node);
+
+        while let Some(top_node) = self.stack.last() {
+            if self.stack.len() > 1 {
+                if let Some(second_top_node) = self.stack.get(self.stack.len() - 2) {
+                    if top_node.height() == second_top_node.height() {
+                        let merged_node = Node::create_node(second_top_node, top_node);
+                        self.stack.pop();
+                        self.stack.pop();
+                        self.stack.push(merged_node);
+                    } else {
+                        break
+                    }
+                }
+            } else {
+                break
+            }
+        }
+    }
+
+    pub fn root(&mut self) -> Bytes32 {
```

Comment:
This should probably consume `self` since you are modifying the internal stack:

```suggestion
    pub fn root(self) -> Bytes32 {
```

Reply:
```rust
pub fn root(mut self) -> Bytes32 {
```
has to be `mut` without `&`